### ISSUE TITLE
`prefer-for-of` should include some Array methods (forEach, reduce)

### ISSUE THREAD

**Issue by fregante** [eslint/eslint#12748]
**Please describe what the rule should do:**

Rewriting `forEach` and `reduce` loops to `for-of` increases their readability.

- `forEach` ~~isn't beneficial to anything unless you're really into fp and one-liners.~~ makes you create a single-use function when you might not need one, especially when nesting it.
- `reduce` tends to create backwards logic, where the starting value of the loop is placed at the end. Example thread: https://twitter.com/jaffathecake/status/1213077702300852224

Other methods like `map` and `some` on the other hand are very effective and readable, so they should be left alone.

Prior art:

- https://github.com/mysticatea/eslint-plugin/blob/master/docs/rules/prefer-for-of.md
- https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-for-of.md

**What category of rule is this? (place an "X" next to just one item)**

[ ] Warns about a potential error (problem)
[x] Suggests an alternate way of doing something (suggestion)
[ ] Enforces code style (layout)
[ ] Other (please specify:)

**Provide 2-3 code examples that this rule will warn about:**

👎 Examples of **incorrect** code for this rule:
```js
[1, 2, 3].forEach(item => {
	console.log(item);
});

const accumulator = [1, 2, 3].reduce((accumulator, item) => {
	accumulator += item;
	return accumulator;
}, 0);
```

👍 Examples of **correct** code for this rule:
```js
for (const item of [1, 2, 3]) {
	console.log(item);
}

let accumulator = 0;
for (const item of [1, 2, 3]) {
	accumulator += item;
}

// async functions are allowed
[1, 2, 3].forEach(async item => {
	throw new Error(item);
});
```

**Why should this rule be included in ESLint (instead of a plugin)?**

You already have [`prefer-for-of`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-for-of.md) in `typescript-eslint` (which shouldn't be there in the first place given it's about a very basic feature of the language and unrelated to types)

**Comment by ljharb** [comment]
We'd need to also support the reverse; the airbnb stylegude forbids all loops and requires use of forEach and reduce for these cases.

**Comment by fregante** [comment]
There's already `no-iterator` for that

**Comment by timdp** [comment]
Note that the Array functions also give you:

- the index of the current element: a `let` outside the loop pollutes the outer scope.

- a reference to the array: allows for composition and reuse of the function.

So I feel like you're dismissing the "unless you're really into fp and one-liners" too easily.

**Comment by sindresorhus** [comment]
> the index of the current element: a let outside the loop pollutes the outer scope.

```js
for (const [index, element] of array.entries()) {}
```

> a reference to the array: allows for composition and reuse of the function.

Example?

**Comment by fregante** [comment]
> allows for composition and reuse of the function.

That's part of "unless you're into fp"

I think this rule _could_ apply only to inline functions to allow reusability of other functions (but it could still be an option)

> a reference to the array

For inline functions it's still possible:

```js
[1, 2, 3].forEach((item, index, array) => {
	console.log(item, 'in', array);
});
```
```js
const array = [1, 2, 3];
for (const item of array) {
	console.log(item, 'in', array);
}
```

**Comment by timdp** [comment]
An example of both would be modifying large arrays in-place. Let's say I have a few large images and I want to invert their colors without allocating new ones. After all, the operation is easily reverted.

```js
image1.pixels.forEach(invertInPlace)
image2.pixels.forEach(invertInPlace)
```

And yes, you could also store the images in an array and loop over them, but that's just another use case. This doesn't feel like something to enforce at the codebase level.

As for `Array#entries()`: works fine in Node, but in older browsers, you'll need a polyfill/ponyfill. Another reason not to enforce it.

**Comment by fregante** [comment]
> As for `Array#entries()`: works fine in Node, but in older browsers, you'll need a polyfill/ponyfill. Another reason not to enforce it.

My understanding is that the existence of a rule in eslint doesn't necessarily mean that it's enforced nor included by `eslint:recommended`. Many use `eslint` just for Node, apps and other environments where `Array#entries()` is available or easily polyfillable.

**Comment by timdp** [comment]
Sure. Just saying there are a lot of reasons why one might _not_ want this rule.

**Comment by fregante** [comment]
> ```js
> image1.pixels.forEach(invertInPlace)
> ```

Now I'm kinda curious how you would `invertInPlace` in there. 😅 

Side note: I'd probably rewrite that as a standalone utility rather than a callback, as I probably would most other examples of `array.forEach(externalFunction)`

```js
invert(image1.pixels);
invert(image2.pixels);
```

**Comment by mdjermanovic** [comment]
> **Why should this rule be included in ESLint (instead of a plugin)?**
> 
> You already have [`prefer-for-of`](https://github.com/typescript-eslint/typescript-eslint/blob/master/packages/eslint-plugin/docs/rules/prefer-for-of.md) in `typescript-eslint` (which shouldn't be there in the first place given it's about a very basic feature of the language and unrelated to types)

To clarify, this is a proposal for a new rule to disallow just `.forEach()` and `.reduce()` in favor of `for-of`?

Asking to avoid confusion because the mentioned `typescript-eslint` rule disallows `for` loops, not `.forEach()` and `.reduce()`.

**Comment by fregante** [comment]
> Asking to avoid confusion because the mentioned `typescript-eslint` rule disallows `for` loops, not `.forEach()` and `.reduce()`.

Correct. This would probably be an extension of that rule, the `prefer-for-of` name applies here as well.

Otherwise this rule would be `avoid-simple-array-looping-methods` or something along those lines.

**Comment by mdjermanovic** [comment]
Then, just to confirm, this is a proposal for a new ESLint core rule which would disallow `.forEach()`, `.reduce()` and `for` loop when any of these can be replaced with a `for-of` loop?

**Comment by fregante** [comment]
Right.

**Comment by ljharb** [comment]
And also the reverse, so the most used JS styleguide can forbid for-of loops in these same scenarios :-)

**Comment by fregante** [comment]
Here are some exceptions where `.forEach` can make sense for all preferences (i.e. the `for-of` alternative is unnecessarily longer):

- `arr.forEach(nonInlineFunction)` (optional?)
- `.forEach(async item => { throw new Error(item) })` (always allowed)

And some examples of bad looping habits involving `.forEach`:

- nested looping, which should trigger [`no-loop-function`](https://eslint.org/docs/rules/no-loop-func) but sadly doesn't:
	```js
	arr.forEach(i => {
		i.forEach(j => {
			log(j)
		})
	})
	```
- loop chains with inline functions:
	```js
	arr
		.map(a => a.wow)
		.filter(wow => wow.isCool)
		.forEach(coolWow => console.log(coolWow))
	```

**Comment by ljharb** [comment]
Throwing an error inside an async function in forEach shouldn’t be allowed; that makes the error impossible to catch. An async function generally shouldn’t be used in any context where the return value is ignored.

**Comment by fregante** [comment]
> that makes the error impossible to catch

Errors in synchronous `setTimeout` callbacks and DOM event handlers can't be caught either. By your logic `setTimeout(function () {})` shouldn't be allowed either.

I was looking at my own usage of `.forEach` and found that `async` exception in a situation where I purposefully wanted to ignore the error while still invoking a list of functions: 

https://github.com/sindresorhus/refined-github/blob/7f1f2c52d8deeb0fddb7a7322af7dd090ac19461/source/libs/on-new-comments.ts#L9-L12

I understand however that this is more of exception though, so perhaps I agree with you.

**Comment by ljharb** [comment]
You don’t need an async function in that example since you’re not using await nor the returned promise; that’s just needless overhead.

**Comment by fregante** [comment]
> You don’t need an async function in that example since you’re not using await nor the returned promise; that’s just needless overhead.

Without `async` the loop will be stopped in case of error. I'm open to alternatives, but currently it's not "just needless overhead"

**Comment by ljharb** [comment]
I suppose it didn't occur to me that an asynchronous function might throw a sync error; that's usually described as "zalgo" - but if that's a concern, and wrapping it in try/catch isn't sufficient, then fair enough.
